r-package/grf/tests/testthat/test_forest_summaries.R

```diff
@@ -267,3 +267,64 @@ test_that("best linear projection works as expected with instrumental forest", {
   expect_equal(blp[, "Estimate"], ate[["estimate"]], tolerance = 1e-10)
   expect_equal(blp[, "Std. Error"], ate[["std.err"]], tolerance = 1e-4)
 })
+
+test_that("ground truth does better than forest-estimates of overlap weights", {
+  # A hard setup
+  n <- 1000
+  p <- 5
+  X <- matrix(rnorm(n * (p)), n, p)
+  eX <- 1 / (1 + exp(-10 * X[, 2]))
+  W <- rbinom(n, 1, eX)
+  M <- X[, 2]
+  TAU <- (1 + X[, 2])^2
+  Y <- M + (W - 0.5) * TAU + rnorm(n)
+
+  W.forest <- regression_forest(X, W)
+  W.hat.forest <- predict(W.forest)$predictions
+  suppressWarnings(W.lm <- glm(W ~ X, family = "binomial"))
+  W.hat.lm <- predict(W.lm, type = "response")
+
+  blp.true <- lm(TAU ~ X, weights = eX * (1 - eX))
+  blp.lm.wts <- lm(TAU ~ X, weights = W.hat.lm * (1 - W.hat.lm))
+  blp.forest.wts <- lm(TAU ~ X, weights = W.hat.forest * (1 - W.hat.forest))
+
+  expect_equal(coef(blp.lm.wts), coef(blp.true), tolerance = 0.1)
+  expect_equal(coef(blp.forest.wts), coef(blp.true), tolerance = 0.5)
+})
+
+test_that("overlap weighted best linear projection works as expected in a simple setup", {
+  n <- 1000
+  p <- 5
+  X <- matrix(rnorm(n * (p)), n, p)
+  eX <- 1 / (1 + exp(-3 * X[, 1]))
+  W <- rbinom(n, 1, eX)
+  TAU <- X[, 2]
+  Y <- X[, 1] + (W - 0.5) * TAU + rnorm(n)
+  true.blp <- lm(TAU ~ X, weights = eX * (1 - eX))
+
+  cf <- causal_forest(X, Y, W)
+  blp.wate <- best_linear_projection(cf, X, target.sample = "overlap")
+  expect_equal(blp.wate[3, 1], coef(true.blp)[[3]], tolerance = 3 * blp.wate[3, 2])
+
+  # TODO these SEs are not smaller than with overlap:
+  #suppressWarnings(blp <- best_linear_projection(cf, X))
+})
```

Comment:
The overlap BLP SEs aren't necessarily always smaller (A4):

```
> blp[, "Std. Error"]
(Intercept)          A1          A2          A3          A4          A5 
 0.21364951  0.24468694  0.09034983  0.19493940  0.07519488  0.24777211 
> blp.wate[, "Std. Error"]
(Intercept)          A1          A2          A3          A4          A5 
 0.09043969  0.10037017  0.08237607  0.09305848  0.08076817  0.08877049 
```

Reply:
I'm not sure what a good test here would be, I thought something like
`se.ratios <- blp.wate[, "Std. Error"] / blp[, "Std. Error"]`
and requiring that most overlap coeffs have smaller SE by doing
`sum(se.ratios <= 0.95)`: is more than 4

But even this doesn't hold for all realizaitons (i.e that sum could be 0)

Reply:
The overlap SE's are on average smaller, though not always, so leaving this simulation assurance here instead of adding a time-consuming test:

```R
res = t(replicate(100, {
  n <- 1000
  p <- 5
  X <- matrix(rnorm(n * (p)), n, p)
  eX <- 1 / (1 + exp(-3 * X[, 1]))
  W <- rbinom(n, 1, eX)
  TAU <- X[, 2]
  Y <- X[, 1] + (W - 0.5) * TAU + rnorm(n)
  true.blp <- lm(TAU ~ X, weights = eX * (1 - eX))

  cf <- causal_forest(X, Y, W)
  blp.wate <- best_linear_projection(cf, X, target.sample = "overlap")
  expect_equal(blp.wate[3, 1], coef(true.blp)[[3]], tolerance = 3 * blp.wate[3, 2])

  blp <- best_linear_projection(cf, X)
  se.ratio = blp.wate[, "Std. Error"] / blp[, "Std. Error"]

  se.ratio
}))
summary(res)
# (Intercept)           A1               A2               A3               A4                A5
# Min.   :0.1276   Min.   :0.1886   Min.   :0.1185   Min.   :0.4008   Min.   :0.06583   Min.   :0.1860
# 1st Qu.:0.7326   1st Qu.:0.8098   1st Qu.:0.7612   1st Qu.:0.7363   1st Qu.:0.79253   1st Qu.:0.7530
# Median :0.8248   Median :0.9386   Median :0.8886   Median :0.8539   Median :0.87241   Median :0.8707
# Mean   :0.8051   Mean   :0.9277   Mean   :0.8609   Mean   :0.8410   Mean   :0.87391   Mean   :0.8537
# 3rd Qu.:0.9118   3rd Qu.:1.0749   3rd Qu.:1.0221   3rd Qu.:0.9594   3rd Qu.:1.01060   3rd Qu.:0.9588
# Max.   :1.1154   Max.   :1.5821   Max.   :1.1848   Max.   :1.1481   Max.   :1.21043   Max.   :1.2538
```